docs/user/implement_algo.md

```diff
@@ -0,0 +1,434 @@
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #0 | Saved
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #0 | Time 0.01 s
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #0 | EpochTime 0.01 s
+-------------  -
+TotalEnvSteps  0
+-------------  -
+It is epoch 1
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #1 | Saving snapshot...
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #1 | Saved
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #1 | Time 0.01 s
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #1 | EpochTime 0.00 s
+-------------  -
+TotalEnvSteps  0
+-------------  -
+It is epoch 2
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #2 | Saving snapshot...
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #2 | Saved
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #2 | Time 0.02 s
+2020-07-22 23:32:34 | [debug_my_algorithm] epoch #2 | EpochTime 0.01 s
+-------------  -
+TotalEnvSteps  0
+-------------  -
+```
+
+Now that we have the basics out of the way, we can start actually doing some
+reinforcement learning.
+
+## Gathering Samples
+
+In the above section, we set up an algorithm, but never actually explored the
+environment at all, as can be seen by `TotalEnvSteps` always being zero.
+
+In order to collect samples from the environment, we can set the `sampler_cls`
+and `policy` fields on our algorithm, and call `runner.obtain_samples()`. We'll
+also need to seed the random number generators used for the experiment.
+
+```py
+from garage.samplers import RaySampler
+
+class SimpleVPG:
+
+    sampler_cls = RaySampler
+
+    def __init__(self, env_spec, policy):
+        self.env_spec = env_spec
+        self.policy = policy
+        self.max_path_length = 200
+
+    def train(self, runner):
+        for epoch in runner.step_epochs():
+            samples = runner.obtain_samples(epoch)
+
+from garage import wrap_experiment
+from garage.envs import PointEnv, GarageEnv
+from garage.experiment import LocalRunner
+from garage.experiment.deterministic import set_seed
+from garage.torch.policies import GaussianMLPPolicy
+
+@wrap_experiment
+def debug_my_algorithm(ctxt):
+    set_seed(100)
+    runner = LocalRunner(ctxt)
+    env = GarageEnv(PointEnv())
+    policy = GaussianMLPPolicy(env.spec)
+    algo = SimpleVPG(policy)
+    runner.setup(algo, env)
+    runner.train(n_epochs=500, batch_size=4000)
+
+debug_my_algorithm()
+```
+
+## Training the Policy with Samples
+
+```eval_rst
+Of course, we'll need to actually use the resulting samples to train our
+policy with PyTorch, TensorFlow or NumPy. In this tutorial, we'll implement an
+extremely simple form of REINFORCE :cite:`williams1992simple` (a.k.a. Vanilla
+Policy Gradient) using PyTorch and TensorFlow. We will also implement a simple
+Cross Entropy Method (CEM) :cite:`rubinstein2004cross` using NumPy.
+```
+
+### PyTorch
+
+```eval_rst
+.. literalinclude:: ../../examples/torch/tutorial_vpg.py
+   :lines: 3-5,10-11,16,24-31,33,40-41,46-48,58-72
```

Comment:
i'm pretty concerned that using the line numbers here, because if these files get changes, this documentation might silently look really wrong. does it work if you include only whole files?

Reply:
It sure works if whole files are included. But in the tutorial, we need to split it into several parts to make it clearer and focusing.

Reply:
So the other choice is copy and paste code blocks except for the final complete file.